source/darwin/dispatch_queue_event_loop.c

```diff
@@ -0,0 +1,635 @@
+}
+
+// Iteration function that scheduled and executed by the Dispatch Queue API
+static void s_run_iteration(void *context) {
+    struct scheduled_service_entry *entry = context;
+    struct dispatch_loop_context *dispatch_queue_context = entry->dispatch_queue_context;
+    s_rlock_dispatch_loop_context(dispatch_queue_context);
+
+    if (!begin_iteration(entry)) {
+        goto iteration_done;
+    }
+
+    struct dispatch_loop *dispatch_loop = entry->dispatch_queue_context->io_dispatch_loop;
+    // swap the cross-thread tasks into task-local data
+    struct aws_linked_list local_cross_thread_tasks;
+    aws_linked_list_init(&local_cross_thread_tasks);
+    s_lock_cross_thread_data(dispatch_loop);
+    dispatch_loop->synced_data.current_thread_id = aws_thread_current_thread_id();
+    dispatch_loop->synced_data.is_executing = true;
+    aws_linked_list_swap_contents(&dispatch_loop->synced_data.cross_thread_tasks, &local_cross_thread_tasks);
+    s_unlock_cross_thread_data(dispatch_loop);
+
+    aws_event_loop_register_tick_start(dispatch_loop->base_loop);
+
+    // run the full iteration here: local cross-thread tasks
+    while (!aws_linked_list_empty(&local_cross_thread_tasks)) {
+        struct aws_linked_list_node *node = aws_linked_list_pop_front(&local_cross_thread_tasks);
+        struct aws_task *task = AWS_CONTAINER_OF(node, struct aws_task, node);
+
+        /* Timestamp 0 is used to denote "now" tasks */
+        if (task->timestamp == 0) {
+            aws_task_scheduler_schedule_now(&dispatch_loop->scheduler, task);
+        } else {
+            aws_task_scheduler_schedule_future(&dispatch_loop->scheduler, task, task->timestamp);
+        }
+    }
+
+    // run all scheduled tasks
+    uint64_t now_ns = 0;
+    aws_event_loop_current_clock_time(dispatch_loop->base_loop, &now_ns);
+    aws_task_scheduler_run_all(&dispatch_loop->scheduler, now_ns);
+    aws_event_loop_register_tick_end(dispatch_loop->base_loop);
+
+    end_iteration(entry);
+
+iteration_done:
+    s_scheduled_service_entry_destroy(entry);
+    s_runlock_dispatch_loop_context(dispatch_queue_context);
+}
+
+/**
+ * Checks if a new iteration task needs to be scheduled, given a target timestamp. If so, submits an iteration task to
+ * dispatch queue and registers the pending execution in the event loop's list of scheduled_services.
+ *
+ * If timestamp==0, the function will always schedule a new iteration as long as the event loop is not suspended.
+ *
+ * The function should be wrapped with dispatch_loop->context->lock & dispatch_loop->synced_data.lock
+ */
+static void s_try_schedule_new_iteration(struct dispatch_loop_context *dispatch_loop_context, uint64_t timestamp) {
+    struct dispatch_loop *dispatch_loop = dispatch_loop_context->io_dispatch_loop;
+    if (!dispatch_loop || dispatch_loop->synced_data.suspended) {
+        return;
+    }
+    if (!s_should_schedule_iteration(&dispatch_loop_context->scheduling_state.scheduled_services, timestamp)) {
```

Comment:
If we always call should_schedule inside try_schedule, then we can probably replace the 
```
if (should_schedule(...))
    try_schedule(...)
```
call  in end_iteration